File: ios/TacticalMaps/Map/MeasureSession.swift
import Foundation
import CoreLocation
import Combine

/// State machine for measure tool. When isActive is true the HUD swaps
/// to MeasureToolbar, map taps add vertices, polyline renders alongside
/// drawings but never gets persisted.
///
/// Distances are haversine via CLLocation.distance(from:), areas use
/// spherical-excess approx which is close enough for tactical-map polygons.
final class MeasureSession: ObservableObject {
    @Published private(set) var isActive: Bool = false
    @Published private(set) var points: [CLLocationCoordinate2D] = []

    func start() {
        isActive = true
        points = []
    }

    func cancel() {
        isActive = false
        points = []
    }

    func addPoint(_ c: CLLocationCoordinate2D) {
        points.append(c)
    }

    func undo() {
        guard !points.isEmpty else { return }
        points.removeLast()
    }

    /// Total haversine distance across all points, in metres.
    var totalDistanceMeters: Double {
        guard points.count >= 2 else { return 0 }
        var total = 0.0
        for i in 0 ..< points.count - 1 {
            let a = CLLocation(latitude: points[i].latitude, longitude: points[i].longitude)
            let b = CLLocation(latitude: points[i + 1].latitude, longitude: points[i + 1].longitude)
            total += a.distance(from: b)
        }
        return total
    }

    /// Bearing of the most recent segment, in degrees (0=N, 90=E).
    var lastBearingDegrees: Double? {
        guard points.count >= 2 else { return nil }
        return Self.bearing(from: points[points.count - 2], to: points[points.count - 1])
    }

    /// Bearing in NATO mils (6400 per circle) for the most recent segment.
    var lastBearingMils: Int? {
        guard let deg = lastBearingDegrees else { return nil }
        return Int(round(deg * 6400 / 360)) % 6400
    }

    /// Enclosed area in sq metres for 3+ points. Polygon is closed
    /// implicitly (last point connects back to first).
    var enclosedAreaSquareMeters: Double? {
        guard points.count >= 3 else { return nil }
        let R: Double = 6_378_137  // mean Earth radius
        var area = 0.0
        for i in 0 ..< points.count {
            let p1 = points[i]
            let p2 = points[(i + 1) % points.count]
            let lat1 = p1.latitude  * .pi / 180
            let lat2 = p2.latitude  * .pi / 180
            let lon1 = p1.longitude * .pi / 180
            let lon2 = p2.longitude * .pi / 180
            area += (lon2 - lon1) * (2 + sin(lat1) + sin(lat2))
        }
        return abs(area * R * R / 2)
    }

    private static func bearing(from a: CLLocationCoordinate2D, to b: CLLocationCoordinate2D) -> Double {
        let φ1 = a.latitude  * .pi / 180
        let φ2 = b.latitude  * .pi / 180
        let Δλ = (b.longitude - a.longitude) * .pi / 180
        let y = sin(Δλ) * cos(φ2)
        let x = cos(φ1) * sin(φ2) - sin(φ1) * cos(φ2) * cos(Δλ)
        let θ = atan2(y, x) * 180 / .pi
        return (θ + 360).truncatingRemainder(dividingBy: 360)
    }
}

/// Formatting helpers for MeasureToolbar + the running HUD.
enum MeasureFormat {
    static func distance(_ m: Double) -> String {
        if m < 1000 { return String(format: "%.0f m", m) }
        if m < 100_000 { return String(format: "%.2f km", m / 1000) }
        return String(format: "%.0f km", m / 1000)
    }

    static func area(_ sqm: Double) -> String {
        if sqm < 10_000 { return String(format: "%.0f m²", sqm) }
        if sqm < 1_000_000 { return String(format: "%.2f ha", sqm / 10_000) }
        return String(format: "%.2f km²", sqm / 1_000_000)
    }
}

File: android/app/src/main/java/com/tacticalmaps/map/MeasureSession.kt
package com.tacticalmaps.map

import android.location.Location
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.roundToInt
import kotlin.math.sin

/**
 * Local state holder for the measure tool. Lives in [MapScreen]'s
 * composition; tap events while [isActive] is true call [addPoint], and
 * the running [points] list is rendered as a dashed polyline.
 *
 * Distances use [Location.distanceBetween] (haversine on WGS84); area
 * uses the spherical-excess approximation, accurate enough for tactical
 * polygons on a tabletop scale.
 */
class MeasureSession {
    var isActive by mutableStateOf(false)
        private set
    val points = mutableStateListOf<Pair<Double, Double>>()  // lat, lng

    fun start() {
        isActive = true
        points.clear()
    }

    fun cancel() {
        isActive = false
        points.clear()
    }

    fun addPoint(lat: Double, lng: Double) {
        points.add(lat to lng)
    }

    fun undo() {
        if (points.isNotEmpty()) points.removeAt(points.lastIndex)
    }

    val totalDistanceMeters: Double
        get() {
            if (points.size < 2) return 0.0
            val out = FloatArray(1)
            var total = 0.0
            for (i in 0 until points.size - 1) {
                Location.distanceBetween(
                    points[i].first, points[i].second,
                    points[i + 1].first, points[i + 1].second,
                    out
                )
                total += out[0]
            }
            return total
        }

    val lastBearingDegrees: Double?
        get() {
            if (points.size < 2) return null
            val a = points[points.size - 2]
            val b = points.last()
            val phi1 = a.first  * PI / 180
            val phi2 = b.first  * PI / 180
            val dLambda = (b.second - a.second) * PI / 180
            val y = sin(dLambda) * cos(phi2)
            val x = cos(phi1) * sin(phi2) - sin(phi1) * cos(phi2) * cos(dLambda)
            val theta = atan2(y, x) * 180 / PI
            return (theta + 360) % 360
        }

    val lastBearingMils: Int?
        get() = lastBearingDegrees?.let { (it * 6400 / 360).roundToInt() % 6400 }

    val enclosedAreaSquareMeters: Double?
        get() {
            if (points.size < 3) return null
            val R = 6_378_137.0
            var area = 0.0
            for (i in points.indices) {
                val p1 = points[i]
                val p2 = points[(i + 1) % points.size]
                val lat1 = p1.first  * PI / 180
                val lat2 = p2.first  * PI / 180
                val lon1 = p1.second * PI / 180
                val lon2 = p2.second * PI / 180
                area += (lon2 - lon1) * (2 + sin(lat1) + sin(lat2))
            }
            return abs(area * R * R / 2)
        }
}

/** Format helpers used by the measure HUD. */
object MeasureFormat {
    fun distance(m: Double): String = when {
        m < 1000 -> "%.0f m".format(m)
        m < 100_000 -> "%.2f km".format(m / 1000)
        else -> "%.0f km".format(m / 1000)
    }

    fun area(sqm: Double): String = when {
        sqm < 10_000 -> "%.0f m²".format(sqm)
        sqm < 1_000_000 -> "%.2f ha".format(sqm / 10_000)
        else -> "%.2f km²".format(sqm / 1_000_000)
    }
}
